engine/action_contact.go

```diff
@@ -4,60 +4,137 @@ import (
 	"context"
 	"errors"
 	"fmt"
+
 	"github.com/bitcoin-sv/go-paymail"
-	"github.com/mrz1836/go-cachestore"
+	"github.com/bitcoin-sv/spv-wallet/engine/utils"
 	"github.com/mrz1836/go-datastore"
 )
 
-func (c *Client) NewContact(ctx context.Context, fullName, paymail, senderPubKey string, opts ...ModelOps) (*Contact, error) {
-	// Check for existing NewRelic transaction
-	ctx = c.GetOrStartTxn(ctx, "new_contact")
+var (
+	ErrInvalidRequesterXpub = errors.New("invalid requester xpub")
+	ErrAddingContactRequest = errors.New("adding contact request failed")
+)
 
-	contact, err := getContact(ctx, fullName, paymail, senderPubKey, c.DefaultModelOptions()...)
+func (c *Client) UpsertContact(ctx context.Context, ctcFName, ctcPaymail, requesterXpub string, opts ...ModelOps) (*Contact, error) {
+	reqXPubID := utils.Hash(requesterXpub)
 
-	if contact != nil {
-		return nil, errors.New("contact already exists")
-	}
+	reqPms, err := c.GetPaymailAddressesByXPubID(ctx, reqXPubID, nil, nil, nil)
 	if err != nil {
 		return nil, err
 	}
+	if len(reqPms) == 0 {
+		return nil, ErrInvalidRequesterXpub
+	}
+
+	reqPm := reqPms[0]
```

Comment:
IMHO we should consider the situation that xpub can have more then one paymail assigned, although we don't do it like that in our examples/references.

Maybe just create a task (for the next sprint) to support multiple paymails attached to the xpub. Because when this choosen paymail is used for requesting a contact at the counterparty side.

Reply:
That's why my initial suggestion was to send chosen paymail every time in payload.

Reply:
yes, but my suggestion was, to allow empty chosen paymail, and if it is empty, and a user has only one paymail, then use this only one. And if he has several, then return an error.
This would make it easier for using in the most common situation, when xpub has single paymail assigned, but also allow for usage in case of multiple paymails for the same xpub.

Reply:
IMO, this doesn't make sense and forces additional logic implementation on the client side to handle these two situations - why not expect paymail from the beginning? 
However, I'll change it. I really want to close this task.

Reply:
Done